cpp/src/arrow/compute/kernels/vector_swizzle_test.cc

```diff
@@ -0,0 +1,781 @@
+            " of inverse_permutation is insufficient to store indices of length " +
+            std::to_string(input_length),
+        InversePermutation(indices, /*max_index=*/0, output_type));
+  }
+}
+
+// ----------------------------------------------------------------------
+// Scatter tests
+//
+// Shorthand notation:
+//
+//   A = Array
+//   C = ChunkedArray
+
+namespace {
+
+Result<Datum> Scatter(const Datum& values, const Datum& indices, int64_t max_index) {
+  ScatterOptions options{max_index};
+  ARROW_ASSIGN_OR_RAISE(Datum result, Scatter(values, indices, options));
+  ValidateOutput(result);
+  return result;
+}
+
+void AssertScatterAAA(const std::shared_ptr<Array>& values,
+                      const std::shared_ptr<Array>& indices, int64_t max_index,
+                      const std::shared_ptr<Array>& expected) {
+  ASSERT_OK_AND_ASSIGN(Datum result, Scatter(values, indices, max_index));
+  AssertDatumsEqual(expected, result);
+}
+
+void AssertScatterCAC(const std::shared_ptr<ChunkedArray>& values,
+                      const std::shared_ptr<Array>& indices, int64_t max_index,
+                      const std::shared_ptr<Array>& expected) {
+  ASSERT_OK_AND_ASSIGN(Datum result, Scatter(values, indices, max_index));
+  ASSERT_TRUE(result.is_chunked_array());
+  ASSERT_OK_AND_ASSIGN(auto result_array, Concatenate(result.chunked_array()->chunks()));
+  AssertDatumsEqual(expected, result_array);
+}
+
+void AssertScatterACC(const std::shared_ptr<Array>& values,
+                      const std::shared_ptr<ChunkedArray>& indices, int64_t max_index,
+                      const std::shared_ptr<Array>& expected) {
+  ASSERT_OK_AND_ASSIGN(Datum result, Scatter(values, indices, max_index));
+  ASSERT_TRUE(result.is_chunked_array());
+  ASSERT_OK_AND_ASSIGN(auto result_array, Concatenate(result.chunked_array()->chunks()));
+  AssertDatumsEqual(expected, result_array);
+}
+
+void AssertScatterCCC(const std::shared_ptr<ChunkedArray>& values,
+                      const std::shared_ptr<ChunkedArray>& indices, int64_t max_index,
+                      const std::shared_ptr<Array>& expected) {
+  ASSERT_OK_AND_ASSIGN(Datum result, Scatter(values, indices, max_index));
+  ASSERT_TRUE(result.is_chunked_array());
+  ASSERT_OK_AND_ASSIGN(auto result_array, Concatenate(result.chunked_array()->chunks()));
+  AssertDatumsEqual(expected, result_array);
+}
+
+void DoTestScatterAAA(const std::shared_ptr<Array>& values,
+                      const std::shared_ptr<Array>& indices, int64_t max_index,
+                      const std::shared_ptr<Array>& expected) {
+  AssertScatterAAA(values, indices, max_index, expected);
+}
+
+/// The following helper functions are based on the invariant:
+/// Scatter([V, V], [I', I''], 2 * (m + 1) - 1) == Concat(E, E)
+///
+/// where
+///   V = values
+///   I = indices
+///   m = max_index
+///   I' = ReplaceWithMask(I, i > m, null)
+///   I'' = ReplaceWithMask(I, i < 0, null) + m + 1
```

Comment:
Why are the `i < 0` and `i > m` conditions necessary?

Reply:
No, they are not. Thank you.

They are introduced in the version that used to allow invalid indices (`< 0` or `> m`). After the change that no more allows these indices and removes those test cases, this complexities are indeed unnecessary.

I'll update.

Reply:
Addressed.